feat(byods): beta for BYoDS

bhabalan opened the pull request:
<!--
Hey there,\
Thank you for taking the time to improve our code! 🙂\
Please let us know why this change is necessary and what testing you have performed. \
This ensures our reviewers understand the impact of your change. \

**IMPORTANT**
FAILING TO FILL OUT THIS TEMPLATE WILL RESULT IN REJECTION OF YOUR PULL REQUEST
This is for compliance purposes with FedRAMP program.
-->

# COMPLETES #< [SPARK-614847](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-614847) >

## This pull request addresses

Merge BYoDS SDK to next

## by making the following changes

cherry-pick all commit from feat/byods into next

### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios were tested

Tested all workflows for BYoDS
Vidcast: https://app.vidcast.io/share/1d36745e-8f3f-4375-ac46-e74911a8d613

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

sreenara committed:
docs(byods): added initial readme file for byods (#3810)

bhabalan committed:
feat(byods): project setup (#3811)

Kesari3008 committed:
feat(byods): Spark 549034 jest setup (#3821)

bhabalan committed:
feat(byods): create the main BYODS class implementation (#3824)

rsarika committed:
feat(byods): added verify jws token api (#3857)

rarajes2 committed:
refactor(byods): add http utils and use it everywhere (#3863)

adhmenon committed:
feat(byods-sdk): added-storage-adapter (#3868)

rarajes2 committed:
feat(byods): add trackingid and user-agent to http request (#3886)

adhmenon committed:
feat(byods-sdk): added-logger-package (#3890)

Co-authored-by: RAM1232 <rajukallasmart@gmail.com>

pagour98 committed:
feat(byods): implementation to handle refreshing dataSource token. (#3870)

Co-authored-by: parv_gour <parv_gour@PAGOUR-M-D8B2>
Co-authored-by: rsarika <95286093+rsarika@users.noreply.github.com>

Parimala032 committed:
fix(byods-docs): added readme file (#3901)

Co-authored-by: Adhwaith Menon <111346225+adhmenon@users.noreply.github.com>

adhmenon committed:
feat(byods): implemented-demo-server (#3907)

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan commented:
@coderabbitai don't review

sreenara reviewed (changes requested):
@bhabalan please check the failures on the coverage and integration test side. 
Also, can we check all the coderabbitai comments and see which ones are valid?

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan committed:
feat(byods): add coverage reports

bhabalan committed:
feat(byods): address coderabbit comments

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan committed:
feat(byods): more coverage

bhabalan committed:
fix(byods): revert tooling change for unit tests

bhabalan commented:
Coverage report
<img width="1300" alt="image" src="https://github.com/user-attachments/assets/28a13d8f-7e4a-411e-b922-8b743d376eaa" />

bhabalan committed:
Merge branch 'next' into feat/byods

sreenara commented on packages/byods/package.json:
```diff
@@ -0,0 +1,123 @@
+    "url": "https://github.com/webex/webex-js-sdk.git",
+    "directory": "packages/byods"
+  },
+  "engines": {
```
Should we have this? It will stop developers from integrating this package into an existing application on a lower engine.

bhabalan replied:
Removed

sreenara commented on packages/byods-demo-server/package.json:
```diff
@@ -0,0 +1,55 @@
+    "dev": "NODE_ENV=development BYODS_ENVIRONMENT=integration npx ts-node ./src",
+    "dev:hot": "nodemon --exec \"npm run dev\" --watch ./src -e ts"
+  },
+  "engines": {
```
same here

bhabalan replied:
Removed

sreenara reviewed (changes requested):
Please update the tests done in the PR description after resolving all conflicts and updating to the latest branch.
Also, is there an API doc for the BYODS SDK? I don't see a tsdocs section. Am I missing it?
Couple of questions in the package.json

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan committed:
feat(byods): remove engine requirements

bhabalan committed:
feat(byods): add tsdoc

bhabalan committed:
feat(byods): add debug config for byods demo server

adhmenon committed:
feat(byods): added-list-api-change (#4095)

bhabalan committed:
feat(byods): update npm script for byods demo server

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan commented:
@sreenara , Updated the description with tests performed, Added tsdoc generation, handled the /dataSource items wrapping and also added debug config for vscode

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan commented:
@coderabbitai resolve comments

bhabalan commented on packages/byods-demo-server/src/views/data-source.hbs:
```diff
@@ -0,0 +1,359 @@
+        });
+      }
+
+      function submitForm(endpoint, payload) {
+        fetch(endpoint, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json'
+          },
+          body: JSON.stringify(payload)
+        })
+          .then(response => response.json())
+          .then(data => {
+            if (data.error) {
+              throw new Error(data.error);
+            }
+            alert(`Data source ${isEdit ? 'updated' : 'added'} successfully`);
+            location.reload();
+          })
+          .catch(error => {
+            alert(`Error ${isEdit ? 'updating' : 'adding'} data source: ${error.message}`);
+          });
+      }
```
Not required as it's a demo server

bhabalan commented on packages/byods-demo-server/src/views/data-source.hbs:
```diff
@@ -0,0 +1,359 @@
+        });
+      }
+
+      function decodeJWSToken(jwsToken) {
+        return JSON.parse(atob(jwsToken.split('.')[1]));
+      }
```
Not required as it's a demo server

bhabalan commented on packages/byods/test/unit/spec/data-source-client/index.ts:
```diff
@@ -0,0 +1,348 @@
+        sub: 'mySubject',
+        aud: 'myAudience',
+      });
+      jest.spyOn(global, 'setInterval').mockImplementation((fn: Function) => {
```
Not required as it's a demo server

bhabalan commented on packages/byods-demo-server/src/views/orgs.hbs:
```diff
@@ -0,0 +1,70 @@
+      document.getElementById('saveServiceAppToken').addEventListener('click', function() {
+        const orgId = document.getElementById('orgId').value;
+        const refreshToken = document.getElementById('refreshToken').value;
+        fetch('/api/token/save', {
+          method: 'POST',
+          headers: { 'Content-Type': 'application/json' },
+          body: JSON.stringify({ orgId, refreshToken })
+        })
```
Not required as it's a demo server

bhabalan commented on packages/byods-demo-server/src/views/orgs.hbs:
```diff
@@ -0,0 +1,70 @@
+        <h2>Save Service App Token</h2>
+        <input type="text" id="orgId" placeholder="Organization ID" />
+        <input type="text" id="refreshToken" placeholder="Refresh Token" />
+        <button id="saveServiceAppToken">Save Service App Token</button>
```
Not required as it's a demo server

bhabalan commented on packages/byods-demo-server/src/views/orgs.hbs:
```diff
@@ -0,0 +1,70 @@
+        const orgId = document.getElementById('orgId').value;
+        const refreshToken = document.getElementById('refreshToken').value;
+        fetch('/api/token/save', {
+          method: 'POST',
+          headers: { 'Content-Type': 'application/json' },
+          body: JSON.stringify({ orgId, refreshToken })
+        })
+          .then(response => response.text())
+          .then(data => {
+            alert(data);
+            if (data.includes('successfully')) {
+              fetch('/api/token/list')
+                .then(response => response.json())
+                .then(tokens => {
+                  const orgsList = document.getElementById('orgsList');
+                  const orgsSection = document.getElementById('orgsSection');
+                  orgsList.innerHTML = '';
+                  if (tokens.length > 0) {
+                    orgsSection.style.display = 'block';
+                    tokens.forEach(token => {
+                      const orgCard = document.createElement('div');
+                      orgCard.className = 'org-card';
+                      orgCard.innerHTML = `
+                        <h3>${token.orgId.slice(0, 10)}...${token.orgId.slice(-10)}</h3>
+                        <button onclick="showDataSource('${token.orgId}')">Show Data Source</button>
+                      `;
+                      orgsList.appendChild(orgCard);
+                    });
+                  } else {
+                    orgsSection.style.display = 'none';
+                  }
+                });
+            }
+          });
+      });
+
+      function showDataSource(orgId) {
+        window.location.href = `/data-source?orgId=${orgId}`;
+      }
+    </script>
```
Not required as it's a demo server

bhabalan commented on packages/byods/src/data-source-client/types.ts:
```diff
@@ -0,0 +1,119 @@
+  /**
+   * The error message associated with the data source update request. It is optional for "active" but required for "disabled".
+   */
+  errorMessage?: string;
```
Not required

bhabalan commented on packages/byods/src/Errors/catalog/ExtendedError.ts:
```diff
@@ -0,0 +1,19 @@
+  constructor(msg: ErrorMessage, type: ERROR_TYPE) {
+    super(msg);
+    this.type = type || ERROR_TYPE.DEFAULT;
+  }
```
Modified to use a different pattern

bhabalan commented on packages/byods/src/data-source-client/index.ts:
```diff
@@ -0,0 +1,261 @@
+        } catch (updateError) {
+          // If there is some error then clear the Interval only if interval is active
+          if (interval) {
+            clearInterval(interval);
+          }
+          log.error(
+            new ExtendedError(
+              `Error while updating dataSource token ${updateError}`,
+              ERROR_TYPE.START_AUTO_REFRESH_ERROR
+            ),
+            {
+              file: BYODS_DATA_SOURCE_CLIENT_MODULE,
+              method: 'startAutoRefresh',
+            }
+          );
+
+          return Promise.reject(
+            new Error(`Error while starting auto-refresh for dataSource token: ${updateError}`)
+          );
+        }
+      }, reducedTokenLifetimeMinutes * 60 * 1000); // Converts minutes to milliseconds.
+
+      return Promise.resolve(interval);
+    } catch (error) {
+      log.error(
+        new ExtendedError(
+          `Error while starting auto-refresh for dataSource token: ${error}`,
+          ERROR_TYPE.START_AUTO_REFRESH_ERROR
+        ),
+        {
+          file: BYODS_DATA_SOURCE_CLIENT_MODULE,
+          method: 'startAutoRefresh',
+        }
+      );
+
+      return Promise.reject(
+        new Error(`Error while starting auto-refresh for dataSource token: ${error}`)
+      );
+    }
+  }
```
Used a different timeout mechanism using signals now

bhabalan commented on packages/byods/src/base-client/index.ts:
```diff
@@ -0,0 +1,202 @@
+    // TODO: Handle refresh token expiration
+
+    return token;
+  }
```
Done

bhabalan commented on packages/byods/src/base-client/index.ts:
```diff
@@ -0,0 +1,202 @@
+        Authorization: `Bearer ${token}`,
+        ...this.headers,
+        ...options.headers,
+      },
```
Done

bhabalan commented on packages/byods-demo-server/src/server.ts:
```diff
@@ -0,0 +1,211 @@
+app.post('/api/data-source/refresh-token/:id', async (req: Request, res: Response) => {
+  try {
+    const {id} = req.params;
+    const {tokenLifetimeMinutes} = req.body;
+    await baseClient.dataSource.scheduleJWSTokenRefresh(id, tokenLifetimeMinutes, () =>
+      crypto.randomUUID()
+    ); // Use UUID for nonce
```
Not required as it's a built-in lib

bhabalan commented on packages/byods-demo-server/src/server.ts:
```diff
@@ -0,0 +1,211 @@
+// **** Core Setup **** //
+const app = express();
+
+let sdk: BYODS.BYODS;
+let baseClient: BYODS.BaseClient;
+
```
Not required as it's a demo server

bhabalan commented on packages/byods/src/byods/index.ts:
```diff
@@ -0,0 +1,135 @@
+     */
+    const parsedEnv = process.env.BYODS_ENVIRONMENT || 'production';
+    let jwksUrl = PRODUCTION_BASE_URL;
+
```
Done

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan committed:
Merge branch 'next' into feat/byods

bhabalan committed:
feat(byods): add build:docs for byods to main package.json
